### PR TITLE
istioctl: add support for creating mixer rules

### PR DESCRIPTION
**Pull request by ayj**
This change adds support for creating mixer rules with istioctl. It
supports "create" and "get" as subcommands to "istioctl mixer
rule". The mixer API server address is set with the
ISTIO_MIXER_API_SERVER environment variable for easy integration into
istio demo and smoke test. The server address can also be set with "-m/--mixer" flag.

Example usage:

    $ export ISTIO_MIXER_API_SERVER=localhost:9094
    $ cat rules.yml
    subject: namespace:ns
    revision: "2022"
    rules:
    - aspects:
      - kind: denials

    $ istioctl mixer rule create global myservice.ns.svc.cluster.local -f rules.yml
    $ istioctl mixer rule get global myservice.ns.svc.cluster.local
    revision: "2022"
    rules:
    - aspects:
      - kind: denials
    subject: namespace:ns

There is much room for improvement in the istioctl and API server space which is not addressed in this PR:

1) Unify mixer and proxy API servers

2) Unify proxy and mixer resource file format so that
"istioctl create -f <file>" can be used to create any
istio resource, e.g. DestinationPolicy, RouteRule, MixerRule.

3) Move API server's gRPC and REST API definitions into istio.io/api
instead of spreading it across multiple component repos.

4) istioctl auth to API servers

### CODE REVIEW

File: cmd/istioctl/mixer.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+	mixerRuleCmd = &cobra.Command{
+		Use:          "rule",
+		Short:        "Istio Mixer Rule configuration",
+		SilenceUsage: true,
+	}
+
+	mixerRuleCreateCmd = &cobra.Command{
+		Use:   "create <scope> <subject>",
+		Short: "Create an Istio Mixer rule",
```
Comment (mandarjog) [inline cmd/istioctl/mixer.go]: Create a Istio Mixer rules document.

This document has multiple rules within in.

File: cmd/istioctl/mixer.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+func mixerRulePath(host, scope, subject string) string {
+	if !strings.HasPrefix(host, "http://") {
+		host = "http://" + host
+	}
+	return fmt.Sprintf("%s/api/v1/scopes/%s/subjects/%s/rules", host, scope, subject)
+}
+
+func mixerRuleCreate(host, scope, subject string, rule []byte) error {
+	fmt.Println(mixerRulePath(host, scope, subject))
```
Comment (mandarjog) [inline cmd/istioctl/mixer.go]: is this `fmt` on purpose here?

File: cmd/istioctl/mixer.go [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+	request, err := http.NewRequest("PUT", mixerRulePath(host, scope, subject), bytes.NewReader(rule))
```
Comment (mandarjog) [inline cmd/istioctl/mixer.go]: use `http.MethodPut`

File: cmd/istioctl/mixer.go (hunk 2 above)
Comment (ayj) [inline cmd/istioctl/mixer.go]: Nope, was leftover from debug - removed.

File: cmd/istioctl/mixer.go (hunk 3 above)
Comment (ayj) [inline cmd/istioctl/mixer.go]: Done

File: cmd/istioctl/mixer.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+func mixerRulePath(host, scope, subject string) string {
+	if !strings.HasPrefix(host, "http://") {
+		host = "http://" + host
+	}
+	return fmt.Sprintf("%s/api/v1/scopes/%s/subjects/%s/rules", host, scope, subject)
+}
+
+func mixerRuleCreate(host, scope, subject string, rule []byte) error {
+	request, err := http.NewRequest(http.MethodPut, mixerRulePath(host, scope, subject), bytes.NewReader(rule))
```
Comment (kyessenov) [inline cmd/istioctl/mixer.go]: the benefit of using `NewRequest` is that you can attach cancellation context...

Comment (ayj) [inline cmd/istioctl/mixer.go]: Is there a better way to send a `put` request, or were you suggesting adding a cancellation context?  http.Client doesn't have a `Put` method.

File: bin/init.sh [hunk 5]
```diff
@@ -17,7 +17,10 @@ ln -sf "$(pwd)/bazel-genfiles/external/io_istio_api/proxy/v1/config/cfg.pb.go" \
   vendor/istio.io/api/proxy/v1/config/cfg.pb.go
 mkdir -p vendor/istio.io/manager/test/grpcecho
 ln -sf "$(pwd)/bazel-genfiles/test/grpcecho/echo.pb.go" \
-  vendor/istio.io/manager/test/grpcecho/
+   vendor/istio.io/manager/test/grpcecho/
+mkdir -p vendor/github.com/googleapis/googleapis/google/rpc
+ln -sf $(pwd)/bazel-genfiles/external/com_github_googleapis_googleapis/google/rpc/*.pb.go \
```
Comment (kyessenov) [inline bin/init.sh]: can you also make sure `bin/gazelle` works with this change and do a fix if it doesn't?

File: cmd/istioctl/mixer.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	if err != nil {
+		return fmt.Errorf("failed creating request: %v", err)
+	}
+	request.Header.Set("Content-Type", "application/yaml")
+
+	var client http.Client
+	resp, err := client.Do(request)
```
Comment (kyessenov) [inline cmd/istioctl/mixer.go]: set request timeouts throughout. default timeout is infinite.

File: cmd/istioctl/mixer.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+		return "", fmt.Errorf("failed sending request: %v", err)
+	}
+	defer resp.Body.Close() // nolint: errcheck
+
+	if resp.StatusCode != http.StatusOK {
+		return "", errors.New(http.StatusText(resp.StatusCode))
+	}
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed processing response: %v", err)
```
Comment (kyessenov) [inline cmd/istioctl/mixer.go]: nit: maybe wrap into a function and use multiline.Prefix?

File: bin/init.sh (hunk 5 above)
Comment (ayj) [inline bin/init.sh]: Missed this - thanks

File: cmd/istioctl/mixer.go (hunk 7 above)
Comment (ayj) [inline cmd/istioctl/mixer.go]: I don't quite understand. Are you suggesting regex to match/extract response body contents?

**Comment by ayj** [comment]
Anyone other comments or is this gtg?

File: cmd/istioctl/mixer.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+// TODO This should come from something like istio.io/api instead of
+// being hand copied from istio.io/mixer.
+type mixerAPIResponse struct {
+	Data   interface{} `json:"data,omitempty"`
+	Status rpc.Status  `json:"status,omitempty"`
+}
+
+const (
+	requestTimeout = time.Second
```
Comment (kyessenov) [inline cmd/istioctl/mixer.go]: 60 seconds?

File: cmd/istioctl/mixer.go [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+	}
+	return fmt.Sprintf("%s/api/v1/scopes/%s/subjects/%s/rules", host, scope, subject)
+}
+
+func mixerRuleCreate(host, scope, subject string, rule []byte) error {
+	request, err := http.NewRequest(http.MethodPut, mixerRulePath(host, scope, subject), bytes.NewReader(rule))
+	if err != nil {
+		return fmt.Errorf("failed creating request: %v", err)
+	}
+	request.Header.Set("Content-Type", "application/yaml")
```
Comment (kyessenov) [inline cmd/istioctl/mixer.go]: Is YAML a good format for client-server exchange? I guess it's to make error messages more useful?

**Comment by kyessenov** [comment]
Let's merge this. I suspect we'll need to do several iterations to make it usable, but it's hard to tell now without a working rate-limit or smth else.

File: cmd/istioctl/mixer.go (hunk 7 above)
Comment (kyessenov) [inline cmd/istioctl/mixer.go]: Just want to avoid repetitions of `fmt.Errorf` with same prefix...

File: cmd/istioctl/mixer.go (hunk 4 above)
Comment (kyessenov) [inline cmd/istioctl/mixer.go]: ah, I misunderstood the original comment. There is `http.Put` method that we should avoid.

File: cmd/istioctl/mixer.go (hunk 9 above)
Comment (ayj) [inline cmd/istioctl/mixer.go]: Probably not, but the mixer API server accepts it and I wanted to pass the rule content through transparently with minimal client processing.

**Pull request merged**